### ISSUE TITLE
Invalid axis for batch normalization

### ISSUE THREAD

**Issue by Lotte1990** [Dobiasd/frugally-deep#206]
Loading the model crashes on "Invalid axis for batch normalization". According to the code, the else statement is "Just to make the compiler happy. In reality, this is never called." Well, in my case it is... See output below. I'm attaching a link to my .h5 and .json files. Please tell me what I'm doing wrong.
https://drive.google.com/file/d/1afQKjJjnWEg_gv4-eKiHUQpnIz1XRU_m
```
Loading json ... done. elapsed time: 3.645327 s
Building model ... done. elapsed time: 61.173442 s
Running test 1 of 1 ... terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid axis for batch normalization.
Aborted (core dumped)
```

**Comment by Dobiasd** [comment]
Thanks for the very good report. Of course, this error should never happen. :grimacing:

I was able to reproduce the issue locally with your model.

The variable values in the failing call of `batch_normalization_layer::apply_impl` are:

- `input.shape().rank()` == `1`
- `axis_` == `3`
- `adjusted_axis` = `7`

I'll investigate and get back to you with the results.

**Comment by Lotte1990** [comment]
Great! Thanks a lot.

**Comment by Dobiasd** [comment]
Can you provide the code that generated the model's architecture? It would help me to create an even more minimal example to pin down what is going on.

**Comment by Lotte1990** [comment]
https://drive.google.com/file/d/18FxLvEnuD1sAxte9NIGnkNf5fStx64mC

**Comment by Dobiasd** [comment]
Perfect! That helped a lot.

I reduced it to a quite minimal example, that produces the same error:

```python
# fdeep_issue_206.py
import tensorflow as tf

base_model = tf.keras.applications.ResNet50(input_shape=(32, 32, 3), weights='imagenet', include_top=False)

x = base_model.get_layer('conv4_block6_out').output
x = tf.keras.layers.AveragePooling2D(2, 2)(x)
x = tf.keras.layers.BatchNormalization()(x)

model = tf.keras.models.Model(inputs=base_model.input, outputs=x)
model.save('fdeep_issue_206.h5', include_optimizer=False)
```

```cpp
// fdeep_issue_206.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_issue_206.json");
}
```

```bash
python3 fdeep_issue_206.py
python3 convert_model.py fdeep_issue_206.h5 fdeep_issue_206.json
g++ -std=c++14 -O3 fdeep_issue_206.cpp -o fdeep_issue_206
./fdeep_issue_206
```

```
Loading json ... done. elapsed time: 0.393757 s
Building model ... done. elapsed time: 0.446118 s
Running test 1 of 1 ... terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid axis for batch normalization.
```

Now I'll try to find out why. :slightly_smiling_face:

**Comment by Dobiasd** [comment]
This issue reveals some architectural hack I implemented some time ago, i.e., deriving a tensor's rank from the `shape5`. This produces a problem for tensors with their first valid dimension being `1`.

Now I'm paying off that technical debt by implementing explicit tracking of the rank. This change affects multiple layer types, but will also make some operations more secure. Currently, all these changes just live in a [branch](https://github.com/Dobiasd/frugally-deep/tree/rankfixed), and there are still some things to fix, but it should be doable. I hope (but no guarantee) to provide a solution by the end of the week.

**Comment by Lotte1990** [comment]
That would be great. Please keep me updated.

**Comment by Dobiasd** [comment]
Wow, this was quite a refactoring ride. :grin:
A lot of things needed to change internally. The user-facing changes, however, are minor, i.e.:
- `fdeep::tensor5` has been renamed to `fdeep::tensor`
- `fdeep::tensor5_pos` has been renamed to `fdeep::tensor_pos`
- `fdeep::shape5` has been renamed to `fdeep::tensor_shape`

---

I just tested your model, and it's working now (on my machine). Please checkout the latest version of frugally-deep from the `master` branch, re-convert your model from `.h5` to `.json`, try again, and let me know about the results.

(I'll create an official release with a new tag once the dust has settled.)

**Comment by Lotte1990** [comment]
Thanks! The model has been loading for approximately 15 minutes now... (1 core @ 100%). Is that also the case on your machine for that .h5 file in my first post? I reconverted the same .h5 to a new .json.
```
Loading json ... done. elapsed time: 3.833852 s
Building model ... done. elapsed time: 59.100476 s
Running test 1 of 1 ...
```

**Comment by Lotte1990** [comment]
30+ minutes and still loading. Is there a way to disable the test? Assuming that is the problem (I don't know)...

**Comment by Dobiasd** [comment]
For me it's a lot faster:

```
Loading json ... done. elapsed time: 0.852204 s
Building model ... done. elapsed time: 8.764934 s
Running test 1 of 1 ... done. elapsed time: 17.769101 s
Loading, constructing, testing of fdeep_issue_206.json took 27.417453 s overall.
```

Please make sure you've compiled with speed optimizations enabled and are not in debug mode: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#why-is-my-prediction-roughly-100-times-slower-in-c-as-in-python

**Comment by Lotte1990** [comment]
Indeed, adding `set(CMAKE_BUILD_TYPE "Release")` to my `CMakeLists.txt` did the trick. `-O3` was added automatically by the default `CMAKE_CXX_FLAGS_RELEASE`. I now get an output tensor like this:

`[[[[[[[ 1.8907, -1.4910], [ 1.7981, -1.3269], ..., [ 1.7055, -1.1629], [ 1.6592, -1.0808]]]]]]]`

In python, I perform an np.argmax operation on the output tensor, which gives me a binary image (as np.argmax returns either index 0 or index 1). Is there an easy way to perform this operation on an fdeep::tensor? In its current shape I cannot convert the tensor to an OpenCV image using the code in your FAQ. It results in the following error:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  invalid buffer size
Aborted (core dumped)
```

**Comment by Dobiasd** [comment]
> In its current shape I cannot convert the tensor to an OpenCV image using the code in your FAQ

Sure you can! :slightly_smiling_face:
The output tensor of your model has a shape of `[240, 320, 2]`. You can convert such an `fdeep::tensor` this into an two-byte-channel openCV image like so:

```cpp
#include <fdeep/fdeep.hpp>
#include <opencv2/opencv.hpp>
int main()
{
    const fdeep::tensor output(fdeep::tensor_shape(240, 320, 2), 0);
    const cv::Mat image(cv::Size(320, 240), CV_8UC2);
    fdeep::tensor_into_bytes(output, image.data, image.rows * image.cols * image.channels());
}
```

To find or implement the equivalent of `np.argmax` for an openCV image is out of scope of a frugally-deep-centered discussion.

**Comment by Dobiasd** [comment]
Here is a quick "argmax" implementation. I've not tested it, but you should get the gist:

```cpp
template <typename T>
cv::Mat argmax(const cv::Mat& mat)
{
    assert(mat.channels() <= 255);
    cv::Mat result(cv::Size(mat.cols, mat.rows), CV_8UC1);
    for (int y = 0; y < mat.rows; ++y)
    {
        for (int x = 0; x < mat.cols; ++x)
        {
            cv::Point pos(x, y);
            int max_channel = 0;
            T max_val = mat.at<T>(pos);
            for (int z = 1; z < mat.channels(); ++z)
            {
                const auto val = mat.at<T>(pos);
                if (val > max_val)
                {
                    max_channel = z;
                }
            }
            result.at<uchar>(pos) = max_channel;
        }
    }
    return result;
}
```

It should not only work for `CV_8UC2` images, but also for `CV_8UC3`, `CV_8UC4`, `CV_16SC2`, `CV_64FC2`, etc.

You can call it like so:
```cpp
const auto binary_image = argmax<uchar>(image);
```

**Comment by Lotte1990** [comment]
My bad, I was using the wrong mat_type (i.e. not CV_8UC2). It's working now. I'm splitting the cv::Mat in 2 slices and using `cv::compare(output(slice1),output(slice0),mask,cv::CMP_GT)` to get the argmax.

**Comment by Dobiasd** [comment]
Ah, nice. That solution is way more elegant than my hand-written one.

**Comment by Dobiasd** [comment]
I guess we can close this issue, right?

**Comment by Lotte1990** [comment]
Yes. Thanks a lot!
